internal/app/cache/cache.go

```diff
@@ -236,7 +239,21 @@ func (c *cache) DeleteRequest(key string, req transport.Request) error {
 		"node_id", req.GetNodeID(),
 		"request_type", req.GetTypeURL(),
 	).Debug(context.Background(), "request deleted")
-	metrics.CacheAddRequestSubscope(c.scope, key).Counter(metrics.CacheDeleteSuccess).Inc(1)
+	metrics.CacheDeleteRequestSubscope(c.scope, key).Counter(metrics.CacheDeleteSuccess).Inc(1)
+	return nil
+}
+
+func (c *cache) DeleteKey(key string) error {
+	c.cacheMu.Lock()
+	defer c.cacheMu.Unlock()
+	metrics.CacheDeleteKeySubscope(c.scope, key).Counter(metrics.CacheDeleteKeyAttempt).Inc(1)
+	_, found := c.cache.Get(key)
+	if !found {
+		metrics.CacheDeleteKeySubscope(c.scope, key).Counter(metrics.CacheDeleteKeyError).Inc(1)
+		return fmt.Errorf("unable to delete entry for nonexistent key: %s", key)
+	}
+	c.cache.Remove(key)
```

Comment:
i think we need to cancel the downstream watches and upstream stream when this happens

Reply:
c.cache.Remove() also runs the onEvicted() func internally, which deletes the downstream watches and upstream stream since it is currently set to onCacheEvicted():
https://github.com/envoyproxy/xds-relay/blob/master/internal/app/orchestrator/orchestrator.go#L381

Reply:
ok thats awesome.